weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CompareEvaluationsPage/sections/ExampleCompareSection/exampleCompareSectionUtil.ts

```diff
@@ -323,9 +302,181 @@ export const useFilteredAggregateRows = (state: EvaluationComparisonState) => {
       outputColumnKeys,
       leafDims,
     };
-  }, [filteredRows, inputColumnKeys, leafDims, outputColumnKeys]);
+  }, [filteredRows, leafDims, outputColumnKeys]);
 };
+
+// Get the table digest used in the dataset of the first evaluation,
+// which prepares a request for actually fetching the table rows later
+async function getPartialTableRequest(
+  evaluations: UseExampleCompareDataParams[0]['summary']['evaluations'],
+  filteredRows: UseExampleCompareDataParams[1],
+  targetIndex: UseExampleCompareDataParams[2],
+  getTraceServerClient: () => TraceServerClient
+) {
+  const targetRow = filteredRows[targetIndex];
+  if (targetRow == null) {
+    return null;
+  }
+  const datasetRef = Object.values(evaluations)[0].datasetRef as string;
+
+  const datasetObjRes = await getTraceServerClient().readBatch({
+    refs: [datasetRef],
+  });
+  if (!datasetObjRes.vals[0]) {
+    console.error('Dataset not found');
+    return null;
+  }
+
+  const rowsRef = datasetObjRes.vals[0].rows;
+  const parsedRowsRef = parseRef(rowsRef) as WeaveObjectRef;
+
+  return {
+    project_id: projectIdFromParts({
+      entity: parsedRowsRef.entityName,
+      project: parsedRowsRef.projectName,
+    }),
+    digest: parsedRowsRef.artifactVersion,
+  };
+}
+
+async function loadRowDataIntoCache(
+  rowDigests: string[],
+  cachedRowData: Record<string, any>,
+  cachedPartialTableRequest: {
+    project_id: string;
+    digest: string;
+  },
+  getTraceServerClient: () => TraceServerClient
+) {
+  const rowsRes = await getTraceServerClient().tableQuery({
+    ...cachedPartialTableRequest,
+    filter: {
+      row_digests: rowDigests,
+    },
+  });
+  for (const row of rowsRes.rows) {
+    cachedRowData[row.digest] = row.val;
+  }
+}
+
+type UseExampleCompareDataParams = Parameters<typeof useExampleCompareData>;
+
+export function useExampleCompareData(
+  state: EvaluationComparisonState,
+  filteredRows: Array<{
+    inputDigest: string;
+  }>,
+  targetIndex: number
+) {
+  const getTraceServerClient = useGetTraceServerClientContext();
+
+  // cache the row data for the current target row and adjacent rows,
+  // this is to allow for fast re-renders during pagination
+  const cachedRowData = useRef<Record<string, any>>({});
+  const cachedPartialTableRequest = useRef<{
+    project_id: string;
+    digest: string;
+  } | null>(null);
+
+  // This is to provide a way to manually control the re-render of the target row
+  const [mutationCounter, setMutationCounter] = useState<number>(0);
+  const increaseCounter = useCallback(() => {
+    setMutationCounter(prev => prev + 1);
+  }, []);
+
+  const [loading, setLoading] = useState<boolean>(false);
+
+  const targetRowValue = useMemo(() => {
+    if (isEmpty(filteredRows)) {
+      return undefined;
+    }
+    const digest = filteredRows[targetIndex].inputDigest;
+    return flattenObjectPreservingWeaveTypes(cachedRowData.current[digest]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
I think when we override this specific rule it makes sense to add a comment to explain exactly why we need to do that.